### PR TITLE
merge blocks using upstream provider

### PR DESCRIPTION
**Pull request by bratpiorka**

### CODE REVIEW

**Review by ldorau — CHANGES_REQUESTED** [review]

File: src/provider/provider_coarse.c [hunk 1]
```diff
@@ -447,15 +458,25 @@ static block_t *free_block_merge_with_prev(struct ravl *all_blocks,
...
+    assert(all_blocks);
     assert(free_blocks);
     assert(block);
 
+    // check if blocks could be merged by the upstream provider
+    enum umf_result_t merge_success = UMF_RESULT_ERROR_UNKNOWN;
     if (block->next && block->next->used == false &&
-        (block->data + block->size == block->next->data)) {
+        (block->data + block->size == block->next->data) &&
+        (block->origin == block->next->origin)) {
```
Comment (ldorau) [inline src/provider/provider_coarse.c:473]: `(block->origin == block->next->origin)` - same as above

**Review by ldorau — CHANGES_REQUESTED** [review]

File: src/provider/provider_coarse.c [hunk 2]
```diff
@@ -411,14 +411,25 @@ static block_t *free_blocks_rm_node(struct ravl *free_blocks,
...
+    umf_memory_provider_handle_t upstream_memory_provider,
+    struct ravl *all_blocks, struct ravl *free_blocks, block_t *block) {
+
+    assert(all_blocks);
+    assert(free_blocks);
     assert(block);
 
+    // check if blocks could be merged by the upstream provider
+    enum umf_result_t merge_success = UMF_RESULT_ERROR_UNKNOWN;
     if (block->prev && block->prev->used == false &&
```
Comment (PatKamin) [inline src/provider/provider_coarse.c:424]: I think that a check for the usage of `block` (`block->used == false`) is missing.

File: src/provider/provider_coarse.c [hunk 3]
```diff
@@ -612,15 +633,25 @@ static bool debug_check(coarse_memory_provider_t *provider) {
         }
 
         // there shouldn't be two adjacent not-used blocks
-        // if they allocs are continuous
+        // if they allocs are continuous and could be merged
         if (block->prev && block->used == false &&
             (block->prev->data + block->prev->size == block->data)) {
-            assert(block->prev->used == true);
+            if (block->prev->used == false) {
```
Comment (PatKamin) [inline src/provider/provider_coarse.c:639]: Merge this if-statement with the upper one.

File: src/provider/provider_coarse.c [hunk 4]
```diff
@@ -447,15 +458,25 @@ static block_t *free_block_merge_with_prev(struct ravl *all_blocks,
...
+    umf_memory_provider_handle_t upstream_memory_provider,
+    struct ravl *all_blocks, struct ravl *free_blocks, block_t *block) {
+
+    assert(all_blocks);
     assert(free_blocks);
     assert(block);
 
+    // check if blocks could be merged by the upstream provider
+    enum umf_result_t merge_success = UMF_RESULT_ERROR_UNKNOWN;
     if (block->next && block->next->used == false &&
```
Comment (PatKamin) [inline src/provider/provider_coarse.c:471]: I think that a check for the usage of block (block->used == false) is missing.

File: src/provider/provider_coarse.c [hunk 5]
```diff
@@ -637,6 +668,9 @@ static bool debug_check(coarse_memory_provider_t *provider) {
             assert((block->data + block->size) <= block->next->data);
         }
 
+        printf("coarse block: %zu, %p, %d\n", block->size, (void *)block->data,
```
[the remark this answers is not in the document]
Comment (PatKamin) [reply src/provider/provider_coarse.c:671]: Also, there is no description on the values, ie.:
```
"coarse block: size=%zu, addr=%p, used=%s", ..., block->used ? "true" : "false"
```

File: src/provider/provider_coarse.c [hunk 6]
```diff
@@ -411,14 +411,25 @@ static block_t *free_blocks_rm_node(struct ravl *free_blocks,
...
+    assert(all_blocks);
+    assert(free_blocks);
     assert(block);
 
+    // check if blocks could be merged by the upstream provider
+    enum umf_result_t merge_success = UMF_RESULT_ERROR_UNKNOWN;
     if (block->prev && block->prev->used == false &&
-        (block->prev->data + block->prev->size == block->data)) {
+        (block->prev->data + block->prev->size == block->data) &&
+        (block->prev->origin == block->origin)) {
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/provider_coarse.c:426]: done

File: src/provider/provider_coarse.c (hunk 4 above)
Comment (bratpiorka) [reply src/provider/provider_coarse.c:472]: added assert

File: src/provider/provider_coarse.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/provider/provider_coarse.c]: this was a debug print added by mistake, removed

File: src/provider/provider_coarse.c (hunk 2 above)
Comment (ldorau) [reply src/provider/provider_coarse.c:425]: OK

File: src/provider/provider_coarse.c [hunk 7]
```diff
@@ -612,15 +634,22 @@ static bool debug_check(coarse_memory_provider_t *provider) {
         }
 
         // there shouldn't be two adjacent not-used blocks
-        // if they allocs are continuous
-        if (block->prev && block->used == false &&
+        // if they allocs are continuous and could be merged
+        if (block->prev && block->prev->used == false && block->used == false &&
             (block->prev->data + block->prev->size == block->data)) {
-            assert(block->prev->used == true);
+            enum umf_result_t merge_success = umfMemoryProviderAllocMerge(
```
Comment (ldorau) [inline src/provider/provider_coarse.c:640]: It should be like in `free_block_merge_with_prev()`:
```
umfMemoryProviderAllocMerge(
            upstream_memory_provider, block->prev->origin->data,
            block->prev->origin->size, block->origin->data,
            block->origin->size);
```

File: src/provider/provider_coarse.c [hunk 8]
```diff
@@ -612,15 +634,22 @@ static bool debug_check(coarse_memory_provider_t *provider) {
...
+                block->prev->size, block->data, block->size);
+            assert(merge_success != UMF_RESULT_SUCCESS);
         }
 
-        if (block->next && block->used == false &&
+        if (block->next && block->next->used == false && block->used == false &&
             (block->data + block->size == block->next->data)) {
-            assert(block->next->used == true);
+
+            enum umf_result_t merge_success = umfMemoryProviderAllocMerge(
```
Comment (ldorau) [inline src/provider/provider_coarse.c:649]: It should be like in `free_block_merge_with_next()`:
```
umfMemoryProviderAllocMerge(
            upstream_memory_provider, block->origin->data, block->origin->size,
            block->next->origin->data, block->next->origin->size);
```

**Review by ldorau — CHANGES_REQUESTED** [review]

**Review by ldorau — APPROVED** [review]

**Pull request merged**